src/runtime/omp/omp_allocator.cpp

```diff
@@ -31,6 +51,62 @@ void *omp_allocator::raw_allocate(size_t min_alignment, size_t size_bytes,
     return raw_allocate(32, size_bytes, hints);
   }
 
+  if(hints.target_numa_node){
+
+#ifdef HIPSYCL_WITH_TARGET_NUMA_NODE_PROPERTY
+    std::vector<size_t> numa_hints = hints.target_numa_node.value();
+
+    size_t max_numa_node = static_cast<std::size_t>(numa_max_node());
+
+    if(numa_hints.empty()) {
+      register_error(__acpp_here(),
+                     error_info{"omp_allocator: at least one node must be " \
+                                "specified in the target_numa_node property",
+                                error_type::invalid_parameter_error});
+      return nullptr;
+    }
+
+    // get the bitmask descibing the available numa nodes
+    struct bitmask *available_bm = numa_get_mems_allowed();
+    struct bitmask *bm = numa_allocate_nodemask();
+
+    for(size_t node_id : numa_hints){
+      // verify that the node requested is available
+      if(numa_bitmask_isbitset(available_bm, node_id) == 0){
+          register_error(__acpp_here(),
+                         error_info{"omp_allocator: the numa node " +
+                                    std::to_string(node_id) +
+                                    " requested is out of range",
+                                    error_type::invalid_parameter_error});
+          return nullptr;
+      }
+      bm = numa_bitmask_setbit(bm, node_id);
+    }
+
+    // numa_alloc_interleaved align at PAGESIZE
+    // so min_alignment requirement is always met
+    void* mem = numa_alloc_interleaved_subset(size_bytes, bm);
+    numa_free_nodemask(bm);
+    numa_free_nodemask(available_bm);
+
+    // return if allocation fails
+    if (!mem)
+      return mem;
+
+    // the allocation size is stored in an allocation map so that it can be
+    // retrieved when calling numafree.
+    using value_type = amap_t::value_type;
```

Comment:
I suspect, allocation map is actually more powerful than what is needed here - e.g. it can efficiently map an address at arbitrary offset within an allocation to the root address.

I think for your purpose, a simple key-value mapping (e.g. `std::unordered_map`) might be sufficient since we only need the map in `free()`, where we would expect the root address to already be given as argument.

However, if in the future your goal is e.g. to be able to retrieve for arbitrary addresses some information - e.g. on which NUMA domain they were allocated - then allocation map will indeed be the better fit.